packages/xstate-react/src/useInterpret.ts

```diff
@@ -0,0 +1,110 @@
+import { useState, useEffect } from 'react';
+import useIsomorphicLayoutEffect from 'use-isomorphic-layout-effect';
+import {
+  interpret,
+  EventObject,
+  StateMachine,
+  State,
+  Interpreter,
+  InterpreterOptions,
+  MachineOptions,
+  Typestate,
+  Observer
+} from 'xstate';
+import { toObserver } from 'xstate/lib/utils';
+import { MaybeLazy } from './types';
+import useConstant from './useConstant';
+import { UseMachineOptions } from './useMachine';
+import { useReactEffectActions } from './useReactEffectActions';
+
+export function useInterpret<
+  TContext,
+  TEvent extends EventObject,
+  TTypestate extends Typestate<TContext> = { value: any; context: TContext }
+>(
+  getMachine: MaybeLazy<StateMachine<TContext, any, TEvent, TTypestate>>,
+  options: Partial<InterpreterOptions> &
+    Partial<UseMachineOptions<TContext, TEvent>> &
+    Partial<MachineOptions<TContext, TEvent>> = {},
+  observerOrListener?:
+    | Observer<State<TContext, TEvent, any, TTypestate>>
+    | ((value: State<TContext, TEvent, any, TTypestate>) => void)
+): Interpreter<TContext, any, TEvent, TTypestate> {
+  const machine = useConstant(() => {
+    return typeof getMachine === 'function' ? getMachine() : getMachine;
+  });
+
+  if (
+    process.env.NODE_ENV !== 'production' &&
+    typeof getMachine !== 'function'
+  ) {
+    const [initialMachine] = useState(machine);
+
+    if (machine !== initialMachine) {
+      console.warn(
+        'Machine given to `useMachine` has changed between renders. This is not supported and might lead to unexpected results.\n' +
+          'Please make sure that you pass the same Machine as argument each time.'
+      );
+    }
+  }
+
+  const {
+    context,
+    guards,
+    actions,
+    activities,
+    services,
+    delays,
+    state: rehydratedState,
+    ...interpreterOptions
+  } = options;
+
+  const service = useConstant(() => {
+    const machineConfig = {
+      context,
+      guards,
+      actions,
+      activities,
+      services,
+      delays
+    };
+    const machineWithConfig = machine.withConfig(machineConfig, {
+      ...machine.context,
+      ...context
+    } as TContext);
+
+    return interpret(machineWithConfig, {
+      deferEvents: true,
+      ...interpreterOptions
+    });
+  });
+
+  useIsomorphicLayoutEffect(() => {
+    if (observerOrListener) {
+      service.subscribe(toObserver(observerOrListener));
+    }
+    service.start(
+      rehydratedState ? (State.create(rehydratedState) as any) : undefined
+    );
+
+    return () => {
+      service.stop();
+    };
+  }, []);
+
+  // Make sure machine options are kept updated when they change.
+  // This mutation assignment is safe because the service instance is only used
+  // in one place -- this hook's caller.
+  useEffect(() => {
+    Object.assign(service.machine.options, {
```

Comment:
This introduces a small behavioral change - previously some of the things were updated using separate `Object.assign` calls, so a particular implementation type was **merged** with what was already held in the machine whereas now things are **swapped out**. 

I'm actually not sure which way is better and I think we already have some related issues to that (albeit for raw Machine APIs and not the React integration). We should establish how things should be handled in v5 - there are also some typing issues related to what we allow right now. I would actually love to discuss this during the upcoming call - let's add this to the agenda. 

While this stays somewhat undecided - or not fully speced~ - I would say that it's better to preserve the current semantics for the time being and possibly change this in the future. So this would become:
```js
Object.assign(service.machine.options.guards, guards)
Object.assign(service.machine.options.actions, actions)
Object.assign(service.machine.options.services, services)
Object.assign(service.machine.options.delays, delays)
```

Oh! And this is especially important because the version in this PR actually **removes** implementation types:
```js
var options = { actions: { doSmth: () => {} } }
Object.assign(options, { actions: undefined })
// oops, options don't have any `actions` now.
```
This definitely deserves to get a regression test as this is a pretty common and important scenario - to not provide any implementations using `useInterpret`/`useMachine` options and only provide them using a fully static definition.